torch/csrc/jit/passes/tensorexpr_fuser.cpp

```diff
@@ -0,0 +1,205 @@
+#include <torch/csrc/jit/pass_manager.h>
+#include <torch/csrc/jit/passes/alias_analysis.h>
+#include <torch/csrc/jit/passes/common_subexpression_elimination.h>
+#include <torch/csrc/jit/passes/dead_code_elimination.h>
+#include <torch/csrc/jit/passes/utils/subgraph_utils.h>
+
+namespace torch {
+namespace jit {
+
+const Symbol& getTensorExprSymbol() {
+  static Symbol s = Symbol::fromQualString("tensorexpr::Group");
+  return s;
+}
+
+value_list sortReverseTopological(
+    ArrayRef<torch::jit::Value*> inputs,
+    torch::jit::Block* block) {
+  value_list result;
+  for (auto i : inputs) {
+    if (i->node()->owningBlock() == block) {
+      result.push_back(i);
+    }
+  }
+  // Sort in reverse topological order
+  std::sort(
+      result.begin(),
+      result.end(),
+      [&](torch::jit::Value* a, torch::jit::Value* b) {
+        return a->node()->isAfter(b->node());
+      });
+  return result;
+}
+
+bool canHandle(Node* node, AliasDb& aliasDb) {
+  // TODO: actually support some ops
+  return false;
+}
+
+#define REQ(cond)                           \
+  if (!(cond)) {                            \
+    GRAPH_DEBUG("Failed cond " #cond "\n"); \
+    return false;                           \
+  }
+
+bool canMerge(Node* consumer, Node* producer, AliasDb& aliasDb) {
+  // Only handle complete tensor types
+  for (torch::jit::Value* output : consumer->outputs()) {
+    REQ(output->isCompleteTensor());
+  }
+
+  // Only fuse within a block
+  REQ(consumer->owningBlock() == producer->owningBlock());
+
+  // Symbolic checks
+  REQ(canHandle(producer, aliasDb));
+  REQ(
+      (canHandle(consumer, aliasDb) ||
+       consumer->kind() == getTensorExprSymbol()));
+
+  // Alias checks
+  REQ(aliasDb.couldMoveAfterTopologically(consumer, producer));
+
+  return true;
+}
+#undef REQ
+
+Node* getOrCreateTensorExprSubgraph(Node* n) {
+  if (n->hasAttribute(attr::Subgraph) && n->kind() == getTensorExprSymbol()) {
+    return n;
+  }
+  return SubgraphUtils::createSingletonSubgraph(n, getTensorExprSymbol());
+}
+
+c10::optional<Node*> tryMerge(
+    Node* consumer,
+    Node* producer,
+    AliasDb& aliasDb) {
+  GRAPH_DEBUG(
+      "Trying producer ",
+      producer->kind().toQualString(),
+      " and consumer ",
+      consumer->kind().toQualString(),
+      ":\n");
+
+  if (!canMerge(consumer, producer, aliasDb)) {
+    return c10::nullopt;
+  }
+
+  consumer = getOrCreateTensorExprSubgraph(consumer);
+
+  aliasDb.moveAfterTopologicallyValid(consumer, producer);
+  SubgraphUtils::mergeNodeIntoSubgraph(producer, consumer);
+
+  return consumer;
+}
+
+std::pair<graph_node_list::iterator, bool> scanNode(
+    Node* consumer,
+    AliasDb& aliasDb) {
+  auto inputs =
+      sortReverseTopological(consumer->inputs(), consumer->owningBlock());
+
+  // Grab the iterator below consumer.  We'll use that to determine
+  // where to resume iteration, even if consumer gets relocated within
+  // the block.
+  auto iter = --consumer->reverseIterator();
```

Comment:
I find the iterator flow here a little difficult to understand—decrementing a reverse iterator, double incrementing an iterator, doing it prefix, etc. Is there some way we can use the `next()`/`prev()` APIs to make this a little clearer?

Reply:
There should be an easier way, yes - e.g. `create_autodiff_subgraph` does a similar thing and doesn't have this decrements/double-increments. I will follow up on this as we start adding actual code.

PS: I also hope this all becomes much simpler once we have functional subgraphs.